File: engine/src/main/java/com/terralite/engine/input/InputState.java
package com.terralite.engine.input;

import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;

public final class InputState {
    private final Set<InputAction> pressed = new LinkedHashSet<>();

    public void press(InputAction action) {
        pressed.add(Objects.requireNonNull(action, "action"));
    }

    public void release(InputAction action) {
        pressed.remove(Objects.requireNonNull(action, "action"));
    }

    public void setPressed(InputAction action, boolean isPressed) {
        if (isPressed) {
            press(action);
        } else {
            release(action);
        }
    }

    public boolean isPressed(InputAction action) {
        return pressed.contains(Objects.requireNonNull(action, "action"));
    }

    public void clear() {
        pressed.clear();
    }

    public Set<InputAction> pressedActions() {
        return Set.copyOf(pressed);
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/input/InputState.kt
package com.terralite.engine.input

class InputState {
    private val pressed: MutableSet<InputAction> = LinkedHashSet()

    fun press(action: InputAction) { pressed += action }
    fun release(action: InputAction) { pressed -= action }

    fun setPressed(action: InputAction, isPressed: Boolean) {
        if (isPressed) press(action) else release(action)
    }

    fun isPressed(action: InputAction): Boolean = action in pressed

    fun clear() { pressed.clear() }

    fun pressedActions(): Set<InputAction> = pressed.toSet()
}
